### PR TITLE
Initial networking support

### PR DESCRIPTION
**Pull request by danielealbano**
This PR implements the initial networking support for cachegrand based on liburing to handle network requests (ie. create the socket server, accept connections, close connections, read/write data, etc).

It's just the networking layer, the supported protocols implementation (redis initially) and the actual code that will glue everything together will come with future.

Because it's just the bare implementation additional functionalities will be added once will be necessary, the implementation may get revised as well in the process.

### CODE REVIEW

**Comment by fireuse** [comment]
I think that using Redis protocol isn't the best choice here. Redis protocol in a text-oriented protocol so it is slower than protocol with binary data. It will also take more memory when parsing. I think that the binary protocol created for this project would be better.

**Comment by danielealbano** [comment]
I totally agree, the REDIS protocol is really terrible from a performance point of view but it gives three huge advantages:
- it would give people the ability to try the platform without having to change anything or just a little;
- we would easily get access to testing & benching suites built for REDIS that can be reused or adapted as needed.;
- it would give us the ability to perform internal testing easily, even just telnet-ing to the server port at the beginning speeding up the development process.

For the longer run I am thinking that having our own protocol based off protocol buffer should be the way to go, I really love the idea to rely on a zero copy implementation in these low level libraries because it would allow us to have a streaming parser, reduce the memory copies to almost none and rely on the state machine to know what to do instead of accumulating the entire request in memory and process it after.

Said that, I don't like too much the idea of using ZMQ, usually these "old" (born a while ago) projects are kind of bloated because they support everything and run everywhere so the internal design has to fight with a million different limitations and can't really take too much advantage of the new possibilities.

The low level library I have in mind to use is called liburing and wraps up the very brand new io_uring interface exposed by the recent kernels (=> 5.1), for the networking is necessary to use a very recent kernel like 5.6 or 5.7.
This interface is so blazing fast that Linus Torsvald is seriously thinking to use it for ALL the syscalls (ie. if you printf something on the console in text mode glibc will endup using io_uring for it).

io_uring is way faster than epoll / kpoll (that is the best implemented by ZMQ), here some benches
https://github.com/frevib/io_uring-echo-server/blob/io-uring-feat-fast-poll/benchmarks/benchmarks.md

image

With 1000 clients there is an improvement between 40% and 70%, it depends on the size of the message, in a caching platform the vast majority of the GET messages are way smaller than 128 bytes so there would be an average speed improvement of 70% on the GET operations (or small commands in general like small values with SET, like INC, SETNX, etc) for free and an average 30/40% speed improvement on the rest (ie big values with SET).

I know that it may seems weird thinking to use something so blazing fast with something so SLOW like the REDIS protocol, but I am planning to build a set of layers to allow us support generic protocols that implement the actions so it will be possible to implement our own and "simulate" 3rd party platforms (redis, memcache, mongodb, http for s3 apis to support hdfs, etc.) without losing a single cpu cycle of speed :)

The idea is to have a set of components:
- a network io layer, currently only io_uring will be implemented so an initial layer will be modelled to take advantage of its features and, the day we will port cachegrand to other OSes, we will be able to swap these layers (pretty sure that a refactoring will be needed but it's better to refactor a layer than starting from scratch)
- a network channels layer, that will take care of handling all the socket (tcp, udp, dtls, xyz) related stuff
- a network protocol layer, that will implement the protocol itself, the needed state machine to support the logic protocol, and therefore the logic to map the commands of the protocol to the internal functions
- a wrapper that puts together the network io, network channel and network protocol layers

So if we decide to start with 2 protocols to show off the performances and have compatibility we will simply have to implement 2 protocol layers.

Also, because it's super early for the project, we don't really need to fully support the protocol just the most common set of commands (get, set, delete, setnx, inc and similar).

This PR is about implementing the network channels from which the network io interface will be forked off, I have only a minimal knowledge of io_uring and liburing so before designing the layer I need to understand which are its need properly although I can easily imagine having a set of non blocking ops like init, listen, connect, read, write, poll, etc.

For reference, the kernel 5.7 and the latest liburing implement a mechanism to automatic select the buffers to use for the IO so polling is not necessary anymore and that's why I need to play with liburing before desining the interface.
Other mechanisms like plain epoll and kqueue/kevent (freebsd) need polling while IOCP in Windows don't.

Said that, I am totally open to discuss every single bit.

**Comment by danielealbano** [comment]
The reason for which I am also a bit picky about the networking library is that with the needed flexibility for the network (almost zero copy kernel space <-> user space, zero copy in our networking stack, etc, etc, etc) and for the threading (ie thread pool using pre-allocated ring buffers cachealigned) we can really let the underlying functionalities shine!

I spent a bit of time to put together this spreadsheet, I will convert it in the next days in markdown page for the benchmarks because I want to re-run them with 50 iterations instead of 10 (it will take more or less 1 entire day).

https://docs.google.com/spreadsheets/d/1wDWsdUAmicuJaoXpgaASTii4o3j1gTWGViW19cV-gv4/edit?usp=sharing

Bottom line: currently the internal hashtable is able to easily scale up to 2048 threads getting almost a x1000 speed improvement and insert 1.6bln of keys in just 715ms.

Of course this doesn't mean that the platform itself will be able to provide these speeds but without any doubt with the needed flexibility for the network and the threading the final product will be able to very fast.

**Comment by danielealbano** [comment]
I am currently investigating as well the usage of SO_REUSEPORT with cpu pinning via an eBPF program or via the SO_INCOMINGCPU socket option, this will simplify greatly all the internal logic to handle the worker threads because each one will handle it's own set of sockets.
